src/cargo/ops/cargo_install.rs

```diff
@@ -220,11 +221,13 @@ fn check_overwrites(dst: &Path,
 
 fn read_crate_list(path: &Path) -> CargoResult<CrateListingV1> {
     let metadata = path.join(".crates.toml");
-    let mut f = match File::open(&metadata) {
-        Ok(f) => f,
-        Err(..) => return Ok(CrateListingV1 { v1: BTreeMap::new() }),
-    };
-    (|| -> CargoResult<_> {
+    let f = File::open(&metadata);
+    if f.as_ref().err().map(|e| e.kind() == io::ErrorKind::NotFound).unwrap_or(false) {
+        return Ok(CrateListingV1 { v1: BTreeMap::new() });
+    }
+    fn fn_once<R, T: FnOnce() -> R>(a: T) -> T { a }
+    fn_once(move || -> CargoResult<_> {
```

Comment:
I think this may be a little more clear if this continues to use `match` and just uses a custom `return Err(io_error.chain_error(...))` to return an error